android/src/main/java/com/oney/WebRTCModule/PermissionUtils.java

```diff
@@ -0,0 +1,133 @@
+import android.os.Build;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.Looper;
+import android.os.ResultReceiver;
+import android.support.v4.app.ActivityCompat;
+import android.support.v4.content.ContextCompat;
+
+
+/**
+ * Helper module for dealing with dynamic permissions, introduced in Android M (API level 23).
+ */
+public class PermissionUtils {
+    /**
+     * Constants for internal fields in the <tt>Bundle</tt> exchanged between the activity
+     * requesting the permissions and the auxiliary activity we spawn for this purpose.
+     */
+    private static final String GRANT_RESULT = "GRANT_RESULT";
+    private static final String PERMISSION = "PERMISSION";
+    private static final String REQUEST_CODE = "REQUEST_CODE";
+    private static final String RESULT_RECEIVER = "RESULT_RECEIVER";
+
+    /**
+     * Incrementing counter for permissions requests. Each request must have a unique numeric code.
+     */
+    private static int requestCode;
+
+    /**
+     * Requests the given permission. The callback will we called with the requested permission and
+     * the granted result. See: https://developer.android.com/reference/android/content/pm/PackageManager.html#PERMISSION_GRANTED
+     * for the result constants.
+     *
+     * @param context Application context / activity.
+     * @param permission Permission which the application is interested in requesting.
+     * @param callback Callback where the results will be reported.
+     */
+    public static void requestPermission(Context context,
+                                         String permission,
+                                         final Callback callback) {
+        ResultReceiver resultReceiver = new ResultReceiver(new Handler(Looper.getMainLooper())) {
+            @Override
+            protected void onReceiveResult (int resultCode, Bundle resultData) {
+                String permission = resultData.getString(PERMISSION);
+                int grantResult = resultData.getInt(GRANT_RESULT);
+
+                callback.invoke(permission, grantResult);
+            }
+        };
+
+        requestCode++;
+
+        Intent permissionsIntent = new Intent(context, PermissionHelperActivity.class);
+        permissionsIntent.putExtra(RESULT_RECEIVER, resultReceiver);
+        permissionsIntent.putExtra(PERMISSION, permission);
+        permissionsIntent.putExtra(REQUEST_CODE, requestCode);
+        permissionsIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+
+        context.startActivity(permissionsIntent);
+    }
+
+    /**
+     * Helper activity for requesting permissions. Android only allows requesting permissions from
+     * an activity and the result is reported in the <tt>onRequestPermissionsResult</tt> method.
+     * Since this package is a library we create an auxiliary activity and communicate back the
+     * results using a <tt>ResultReceiver</tt>.
+     */
+    public static class PermissionHelperActivity extends Activity {
```

Comment:
`PermissionHelperActivity` seems very invasive to me because it "replaces" the `Activity` which effectively requests the permissions by making the `getUserMedia`. Google mentions a "headless fragment" which attaches to an `Activity` and is able to request runtime permissions. Given that `WebRTCModule` appears in source code to be able to give us a `currentActivity` to which it is attached, have "headless fragments" been explored as an alternative to `PermissionHelperActivity`?